flexvolume: read a response from stdout

avagin opened the pull request:
Currently we call cmd.CombinedOutput to read responses, but this method
reads data from stderr and stdout.

It isn't a common practice. Usually useful data are got from stdout
and stderr is used to output error messages or diagnostics.

Messages from stderr will be printed in the kubelet log, so it allows us
to add some diagnostics in plugins.

Signed-off-by: Andrei Vagin <avagin@gmail.com>

```release-note
flexvolume: read responses from stdout and use stderr for error and diagnostic messages
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44459)
<!-- Reviewable:end -->

grodrigues3 commented:
@k8s-bot ok to test

avagin committed:
flexvolume: read a response from stdout

Currently we call cmd.CombinedOutput to read responses, but this method
reads data from stderr and stdout.

It isn't a common practice. Usually useful data are got from stdout
and stderr is used to output error messages or diagnostics.

Messages from stderr will be printed in the kubelet log, so it allows us
to add some diagnostics in plugins.

Signed-off-by: Andrei Vagin <avagin@gmail.com>

avagin commented:
Cc: @chakri-nelluri

avagin commented:
@dchen1107 Could you review this patch?

mtaufen commented on pkg/volume/flexvolume/driver-call.go:
```diff
@@ -124,7 +128,35 @@ func (dc *DriverCall) Run() (*DriverStatus, error) {
 		defer timer.Stop()
 	}
 
-	output, execErr := cmd.CombinedOutput()
```
Would it be possible to encapsulate your new stuff in a version of CombinedOutput that returns two byte sequences? This looks like something that would be generally useful.

avagin replied:
@mtanino Yes, it is possible. In this case we will need to wait a process to use its stderr output. My idea was to log data from stderr as soon as we recieve them. It is usefull when a target process is running for a long time or if it hangs for some reason. What do you think about that?

chakri-nelluri commented:
@avagin I agree with the unix semantics. But today, we log errors irrespective of whether the error is redirected to stdout or stderr. Anything specific that would benefit because of this change?

avagin commented:
@chakri-nelluri Currently we get a combined output from stderr and stdout and then parse it as a json,  so a driver can't use stderr for error or debug message. I suggest to get a response only from stdout and log data from stderr to kubernetes log. I found this very usefull, when I was developing my driver.

```
	output, execErr := cmd.CombinedOutput(
        ...
	status, err := handleCmdResponse(dc.Command, output)

```

If you look at examples, there stderr and stdout of all commands are redirected to /dev/null and it looks ugly:
https://github.com/kubernetes/kubernetes/blob/master/examples/volumes/flexvolume/lvm#L101

With my patch, we will not need this hack.

chakri-nelluri commented:
@avagin This is going to break existing plugins, if we move to stdout for all "JSON" outputs & use stderr for other errors. It is currently assumed all errors message will be packed into "JSON" output and is written to stderr.
I typically log to syslog or log file for all other specific logs.

avagin commented:
https://github.com/kubernetes/kubernetes/issues/30618

avagin commented:
@chakri-nelluri I believe that the most part of users use stdout to print responses and flexvolume is an alfa feature, so it isn't to late to change it to be more developer-friendly. I'm agree that we can use syslog or a file to save debug messages from these plugins, but it looks like a hack and I think the kube log is more suitable place for them. And it is a unix way, so a new developer will not need to guees that stderr doesn't have to be used in a flex plugins.
There are not only me who found this issue #30618. I don't want to argue much about this code. I found some inconsitance and decided to fix it. If you think it is to late for such changes, let me know. Thank you.

chakri-nelluri commented:
@avagin I think there is some confusion about STDOUT & STDERR.

Today's flex volume implementation assumes (Look into the nfs example)
 - If flex volume calls are successful, log JSON output to STDOUT
 - If flex volume calls fail, log JSON output to STDERR and that way you can interpret and log the message.
I do not see how we are breaking the UNIX semantics.

All plugin specific logging is up to the plugin. We want only a meaningful short error message in the response. My 2 cents, we do not want to do a blob read of all the plugin errors and print in Kubelet log.

avagin commented:
> Today's flex volume implementation assumes (Look into the nfs example)

The implementations of flexvolume assumes that it doesn't matter to which descriptor a response  is written, because the caller gets a combined output from both descriptors. The unix-way is to use a power of two channels, rather than merging them into one. The idea to write the same type of responses into stdout and stderr brings nothing new in compare with the case when only one descriptor is used for responses. If we would have used only stdout for responses, the flexvolume interface didn't become less functional.

@chakri-nelluri Thank you for the feedback. I think our discussion moved to a dispute about tastes, so I think we can close this pull-request.

dmission commented:
> Today's  flex volume implementation assumes (Look into the nfs example)
> 
> - If flex volume calls are successful, log JSON output to STDOUT
> - If flex volume calls fail, log JSON output to STDERR and that way you can interpret and log the error message.

This is not enforced by the code. Thus, implementations could break this assumption.

Current approach leads to additional development efforts in drivers:
- A developer should care that both stdout and stderr contain JSON-formatted output
- A developer should write own logging instead of using kubelet's one (and a user of this driver should be taught where to see errors for troubleshooting)
